### PR TITLE
Implement Svelte UI

### PR DESCRIPTION
**Pull request by itsonlyjames**
Alright... I went above and beyond the last few days; I got super invested in this and I'm incredibly happy with where I've gotten it to. 

This work truly sets us up for a lot of fun handling animations and transitions. I have also **completely** rewritten the event handling structure so now it correctly attaches only a single listener to each leaf, and doesn't listen to scrolls for the entire window. 

Switch to source mode and source view and check it out ;)

It updates headers across multiples of the same files.

It doesn't differentiate source/preview for scrolling, everything is housed locally. 

It no longer checks for headings on every scroll, only ones currently in view, and the resolve handles updating of headers. Overalll, it's a **lot** more performant. 

The only thing it requires for it to be exact with the existing version is to run the `trivial` and `calcIndentLevel` funcs—I just wanted to get your eyes over this because it's a _complete_ rewrite of the existing plugin now using Svelte. 

Version 2.0 here we come 😎

### CODE REVIEW

File: src/plugin.ts [hunk 1]
```diff
@@ -224,11 +229,12 @@ export default class StickyHeadingsPlugin extends Plugin {
...
-        ...heading,
-        title: await parseMarkdown(heading.heading, this.app),
-        offset: view.editor.cm.lineBlockAt(heading.position.start.offset),
-      }))
+      getShownHeadings(headings, view)
+        .map(async ({heading, offset}) => ({
+          ...heading,
+          title: await parseMarkdown(heading.heading, this.app),
+          offset,
+        }))
```
Comment (zhouhua) [inline src/plugin.ts:237]: In your code, it seems to assume that the same title has the same offset in both preview mode and source mode (if I understand correctly). However, from what I've observed, there are significant differences. So, I used yesterday's code to correct this part.

File: src/plugin.ts [hunk 2]
```diff
@@ -141,27 +134,39 @@ export default class StickyHeadingsPlugin extends Plugin {
...
 
-  setHeadingsInView(scrollTop: number, item: FileResolveEntry) {
+  async setHeadingsInView(scroller: HTMLElement, item: FileResolveEntry) {
+    const scrollTop = scroller.scrollTop;
+    // fixme: Use a more appropriate method to get the component height.
+    const stickHeaderHeight = scroller.closest('.view-content')?.querySelector<HTMLElement>('.sticky-headings-root')?.clientHeight || 0;
     if (item) {
-      const headingsInView = item.headings.filter(
-        (heading) => heading.offset.top < scrollTop
+      const headings = await this.retrieveHeadings(item.file, item.view);
```
Comment (zhouhua) [inline src/plugin.ts:156]: I understand that during scrolling, you only want to calculate which titles need to be displayed without rechecking all the titles. However, in my testing scenarios, the height calculation in preview mode might not be accurate right when the file is opened. So for now, I'll execute this logic once more. We can look into optimizing it in the future.

**Comment by zhouhua** [comment]
Great job. I carefully reviewed your code and added some of the refactored code from yesterday. It is currently running well on my end.

**Comment by zhouhua** [comment]
I think there are still some minor issues that need to be addressed. I believe just listening to `file-open` is not enough. Currently, there are problems when switching documents within a leaf, opening documents in a new tab, and it doesn't support the hover editor plugin either.

File: src/plugin.ts (hunk 1 above)
Comment (itsonlyjames) [reply src/plugin.ts:237]: Brilliant

File: src/plugin.ts (hunk 2 above)
Comment (itsonlyjames) [reply src/plugin.ts:156]: Maybe we can debounce it?

**Comment by itsonlyjames** [comment]
Agreed. The PR was in draft stage due to not being 100% so merging it to master was preemptive. I've opened https://github.com/zhouhua/obsidian-sticky-headings/pull/27. Please don't merge until I switch it to ready to review.

**Pull request merged**